### ISSUE TITLE
Guidelines for bringing back visual scripting to Godot in the future

### ISSUE THREAD

**Issue by aaronfranke** [godotengine/godot-proposals#8873]
### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

Godot 3.x had visual scripting, but Godot 4.0 removed this. This is ultimately a good thing, because it would be better to bring it back as a GDExtension in the future.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Provide an official visual scripting system. See below for details.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I am here to propose an 8-part plan to bring back visual scripting to Godot to ensure the system is broadly useful in lots of ways, including GLTF interoperability, extensibility, and of course, ease of use. These 8 parts are guiding principles as opposed to steps that can be taken one at a time.

## 1. Not a part of the engine

Visual scripting should be done with GDExtension. It should not be a part of the engine for several reasons:

* Bloat: Many users do not want visual scripting. Especially considering we may want many high-level blocks, it's better to not burden the main repo with all of that custom behavior.
* Extensibility: Of course we will want to allow writing custom script blocks in GDScript, but users may want to customize the visual scripting system's C++ code. This is much easier if it's done in GDExtension.
* Independent development: This will allow the visual scripting system to be developed and updated independently of the engine. This way users don't need to wait for newer Godot versions to get new visual scripting features.

## 2. Speed is not vital

Of course, we don't want to make the visual scripting system be purposefully slow, so we should write it in a language like C++ or possibly Rust. But we need to ensure the visual scripting system is primarily high code quality, not high performance.

Depending on the operations you do, GDScript is about 5 to 20 times slower than C++, or about 2 to 10 times slower than C#. Big Godot projects may have hundreds of GDScript files, some of them may have hundreds or thousands of lines. However, GDScript is still fast enough for most users. Visual scripting would be used with much simpler scripts than GDScript, so it doesn't need to be as fast as GDScript.

Knowing this, even if visual scripting was 100 times slower than C++, that is still plenty fast. We should primarily target code quality and easy extensibility over performance.

## 3. Allow extending in GDScript

Visual scripting needs to be easy to extend. On a 2-person team of 1 artist and 1 high-level GDScript programmer, the GDScript programmer may wish to provide APIs for visual scripting to use.

Therefore we need to ensure it's easy to extend. We need to allow extending visual script block classes with virtual callback methods. We should also ensure the internal code quality is high, to allow users to extend the C++ code if needed.

Still, allowing extending in GDScript should not come at the cost of the performance of built-in script blocks. Having to allocate a Godot `Object` for every script block to allow exposing it would be quite slow and memory-intensive at runtime. Instead, all of the built-in blocks should be use lightweight structs to store the data internally, and we have a more complex "extension" block class that uses Godot `Object` (or possibly `RefCounted`?) to allow extending in GDScript. This does not hurt code quality while being a big win for performance.

## 4. Provide entire Godot API

One of the big problems with VisualScript in Godot 3.x is that it was missing some critical functionality, like setting values in Arrays. All operations available in GDScript should also be available in visual scripting, although not always exactly the same (ex: since visual script blocks can have multiple outputs `for item in array:` visual script blocks can provide both the index and the item).

Aside from that, we should ensure that every API call available in Godot is accessible via the visual scripting system. Every method needs to be callable, every property needs to be accessible, every class needs to be instantiable, etc.

## 5. Provide high-level blocks

In order for visual scripting to be useful for users to write complex scripts without their canvas turning into a mess of blocks and wires, we need to provide high-level script blocks. These are concepts that may not even make sense to have in GDScript.

For example, let's say we want a high-level script block to get the material properties of an object. The material properties may be located on the MeshInstance3D node, or on the Mesh resource contained in it. In GDScript you would have to check if the MeshInstance3D node is valid, check if it has material properties, check if it has a Mesh resource, check if that has material properties, etc. All that is just to find the material, then you would need to check if it's a BaseMaterial3D and grab each of the properties you need from it, like albedo color, roughness, etc. The high-level visual script block would be much simpler: It takes in an object, and it has many output ports for each of the properties on the material. This does not make much sense in GDScript because on the level of GDScript you really do care about the fine details like where the material is located and what type the object is. It also does not make much sense in GDScript because all of those output ports would simply not work, GDScript functions do not have multiple return values and even if they did that would severely clutter your script with unused variables.

## 6. Learn from existing successes

A visual scripting system would be incomplete without getting experienced Unreal Blueprints users to participate in its development. Tons of people have a great time using Unreal Blueprints, we need to ensure we have all of the features that Unreal users are currently using to make successful AAA games using Blueprints (that doesn't imply that visual scripting will enable AAA games of course, just that your choice of scripting language should not limit you).

## 7. Interoperability with GLTF interactivity

The Khronos group is working on an engine-agnostic scripting system in GLTF. This is super cool. Imagine being able to export a script from Godot and import that into Unreal, or export a script from Unreal and import that into Godot. Or imagine using Godot as a "game asset SDK" where you can write a simple behavior script in Godot and that can be imported at runtime into a "metaverse" style game, such as V-Sekai or The Mirror, two platforms made in Godot, without the security and portability concerns of just sending a Godot scene file with GDScript in it.

I think it would be a great idea to design our future visual scripting system to be compatible with GLTF interactivity to the greatest extent possible. This includes both import and export.

Of course, the Godot visual scripting system will need to be a superset of GLTF interactivity, as there will be many concepts in Godot that can't be available in an engine-agnostic format, and we will probably need to write code to "compile down" high-level blocks into simpler ones. But still, we should be able to export most simple behavior scripts, or most features of complex scripts, or import scripts from GLTF into Godot. This is good enough to cover lots of use cases. For complex scripts being brought between engines, if 10% of the visual script blocks do not survive the trip, that still means 90% did survive, and then it's easier to manually fill in the gaps instead of starting from scratch.

To be clear, the below image has nothing to do with what the proposed UI will look like. This is just an abstract representation of the interactivity graph data.

<img width="1210" alt="gltf_interactivity" src="https://github.com/godotengine/godot-proposals/assets/1646875/c48e5133-349e-4237-8076-25c31d96da24">

## 8. Timeline

There is not really any rush in completing this, and in fact, it may be better to wait for several reasons.

For the goal of interoperability with Khronos GLTF interactivity, it would be helpful to wait for the Khronos group to finish standardizing that so we have a good base standard to structure visual scripting around.

For making an editor, @Geometror has been refactoring the GraphEdit system over the past year or so. It would be helpful to wait to work on the visual scripting system until the GraphEdit refactor is done, so that we will not need to redo the editor code to use the new GraphEdit APIs.

There are still other areas that are a much higher priority for engine contributors to work on. Of course since this will be done in GDExtension, the development cycle is not tied to the engine, but this will very likely be done by existing engine contributors (as evidenced by few people stepping up to maintain VisualScript in Godot 3.x). I would rather the engine contributors spend their time for now on important bug fixes than chase after a new shiny thing. I am personally interested in working on this, and I have lots of experience with developing a visual scripting system thanks to my time at The Mirror, but it is not a high priority for me.

I would expect that we can expect this visual scripting system to be completed sometime in the late 2020s (remember it is already 2024), but it is hard to say if that will be 2026 or 2029. Regardless I want to ensure we take the time do this right, and I have no plans to start working on this at all in 2024 due to all of the above reasons.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

This *should* be an add-on in the asset library! An officially endorsed one.

**Comment by synalice** [comment]
While I do think `5. Provide high-level blocks` is a good idea whose absence would make using visual scripting unbearable, there absolutely must be a way to convert a visual script into a `.gd` file. The same way as we have the ability to convert visual shaders or materials into shader code.

Ensuring this will automatically solve `7. Interoperability with GLTF interactivity` since you'd only need to support `GDScript -> GLTF` conversion _(assuming I correctly understood what this GLTF thing is)_.

**Comment by aaronfranke** [comment]
@synalice GDScript -> GLTF conversion would be vastly worse than visual script -> GLTF conversion.

**Comment by octanejohn** [comment]
@passivestar maybe related to your idea for universal geo nodes see gltf section, i know this is visual scripting but its simular

**Comment by AThousandShips** [comment]
One consideration would be to use the proposed IR directly instead of translating into GDScript
* https://github.com/godotengine/godot-proposals/issues/8605

**Comment by passivestar** [comment]
> Unreal is of course not the only engine with a visual scripting system, but it is the most popular. We can learn from the successes of visual scripting in other engines as well.

I used both Blueprints and Unity UVS, and I urge whoever designs this to take a look at Unity. One of my biggest problems with Blueprints was that it was impossible to create custom nodes that have **_persistent state_** (not without recompiling the engine)

Some built-in nodes have state, like for example the "Flip Flop" node that has a counter inside so that it can execute another branch next time it has control, but that's it. Being able to easily create your own nodes with state allows you to describe more complex behaviours in much less nodes. For example I was able to create a "Transition" node that outputs a value that lags behind of its input, which is very useful for a lot of things

**Comment by aaronfranke** [comment]
@passivestar Great suggestion, but I see no reason to prevent making a visual script block with state, so you'll be able to.

**Comment by FR3NKD** [comment]
![visual scripting](https://github.com/user-attachments/assets/00f342f6-6959-49b0-baf9-4c33bbc889b8)
### I believe I'm a perfect candidate for being a visual scripting user let's see how a visual solution can solve most coding problems for a beginner like me:

### Problem 1: Knowing the syntax

Nodes are just a disaster for readability in my opinion and it's time to think about visual scripting differently so forget about nodes, visual scripting is created using vertical **blocks** that stack on top of each other and can be collapsed or expanded, basically just like code but without needing to know the syntax, this would allow you to easily follow coding tutorials as a side effect.

### Problem 2: Arbitrary knowledge required to use the engine classes / General discoverability of API's functions

A big ADD button would suggest new context-relevant blocks when pressed, blocks should be separated by categories and have icons similar to Unity components.

Blocks should be directly translated into GDScript text, you should be able to follow a GDScript tutorial with this visual solution even if it uses normal code, this way all GDScript tutorials would still be relevant and users might be able to learn GDScript.

High level blocks should be an option and not the norm, I should be able to work 100% like in GDScript or use blocks already created for me which are more high level and simple to use. 
Creating these custom blocks should be easy either via blocks or directly with GDScript.
I can imagine a complete novice using only high-level blocks created by others, or an advanced user who assembles a script quickly by simply adding high-level blocks created previously.

A sidebar should show all available blocks divided by category and a small section should recommend blocks relevant to the current context, blocks can be saved to favorites for quick access.

![components](https://github.com/user-attachments/assets/00464456-9598-4d43-98e6-9987af4a94c6)

### Problem 3: Good design patterns

Not being a coder I don't have a clear solution to this problem but this visual language could direct you to somehow use the right design patterns.

### How to do it?
My original idea was to interpret and convert pure GDScript code into a user interface, giving an interface to the code like modern GUIs have given an interface to OSes.

This should have 0 performance impact compared to GDScript because this conversion does not happen at runtime, so the final output is just GDScript code.

In my deepest fantasies this language can also be used by some medium level programmers to be more efficient and type less.

### The most important part of this idea
This idea could be technically infeasible but the most important part of this idea it's undeniable and is the synergy between GDScript and the visual language, if you have to keep only one thing from my vision, let it be this!
Users could use the GDScript documentation, watch existing tutorials, get help from their colleagues who know GDScript, learn the language without immediately having to write code, the benefits are endless.

**Comment by passivestar** [comment]
> Nodes are just a disaster for readability

Not sure about that. When you have branching, nodes do a better job at visualizing the execution flow than text or stacked blocks

Also being able to see in real time what graph branches run and what values are getting passed is convenient for debugging because you can see at a glance if some value is off or if some branch doesn't get reached

https://github.com/user-attachments/assets/98527e65-caf8-40f1-ba41-3f0ecd5cc73f

**Comment by FR3NKD** [comment]
This kind of visualization is admittedly pretty cool (and useful). I just think spaghetti are unreadable and it is very easy to create spaghetti with nodes.

**Comment by passivestar** [comment]
It's about as easy to create spaghetti code in GDScript. It's just less obvious that you've created it 🙂

I've made a lot of blender tools in both python and geonodes. Not making a mess with nodes is easier than it seems. All of my graphs looks like this no matter the complexity, I don't have long noodles anywhere:

<img width="1279" alt="image" src="https://github.com/user-attachments/assets/77253533-74b5-4eca-9ecb-490fb5b90434">

**Comment by FR3NKD** [comment]
I use nodes out of necessity (for shading and stuff) but I have always found them intimidating and visual scripting should be as intuitive as possible.

I also never liked how you have to zoom out to see the whole graph but then you can't read the text of the nodes.

I don't know.. I don't love them, but this is just my personal preference.

![blueprints](https://github.com/user-attachments/assets/22f459e8-ade3-4844-9575-e076fb257e6f)

**Comment by FR3NKD** [comment]
I would like to add that we had node based visual scripting in Godot and today we have Orchestrator but It doesn't seem to be that popular, why not try something different? I don't think nodes should be the default option necessarily
![visual scripting](https://github.com/user-attachments/assets/b97cb811-b30a-4a24-8540-aefd49c64347)

**Comment by aaronfranke** [comment]
@FR3NKD You are welcome to try something different by developing an addon like what you want yourself. Godot is extensible, you can plug in new scripting languages, for example Orchestrator does this.

**Comment by FR3NKD** [comment]
I'm not a programmer, that's why I'm trying to contribute to this with a non-programmer point of view.

When you create a tools you should put yourself in the shoes of the end user, this can be difficult if you are an expert in the field you ar trying to make accessible.

We should seek feedback from people who understands programming while not actually knowing how to code (they are rare 😅).

I admit that my opinion about nodes vs blocks can be subjective (although I suspect many feel the same way) so you are welcome to ignore that but don't dismiss my feedback all together 😁

**Comment by NathanLovato** [comment]
In terms of user interface and visualizing the visual script, you don't necessarily have to think in terms of _either_ block-based or graph-based visual languages. They're two representations of the logic, and you can combine both in the same visual language as Game Maker Studio 2 does.

You could also consider the visual interface as views of the same underlying logic, starting with one view for development efficiency and adding more as needed.

Graphs are really good at representing pipelines, which is why they're the go-to in technical art - it's a lot of pipelining, some branching, and merging back to one to a few outputs. They're also good for branching to many outputs, like in branching dialogues.

They're definitely prone to things looking entangled and overlapping. Typically in game logic, when you have variables used by many nodes in the graph, and the variables are represented as input nodes, you get a lot of lines crossing each other and it's difficult to organize (you can imagine UI solutions for that like modal views of a single branch of the graph, hiding away the complexity, but I've yet to see something like this).

Block-based representations are really good for logic, and for your typical imperative gameplay programming. Especially as the code grows. They usually promote an imperative/procedural style and work great for this. Basically the same as text: when the code mostly reads top to bottom, it works great.

You could also mix the two similarly or differently from game maker, allowing for example users to plug blocks inside of nodes in a graph, keeping imperative logic vertical, reading like a book, and collapsible. There's certainly a lot to think about in terms of integration with Godot's specific UX and fundamental building blocks.

Anyway, that's just to point out that graphs and blocks could be thought of as views on logic and not as an either/or upfront choice. Under the hood, they can use a node tree that gets converted to bytecode all the same.

As visual languages are all about user interface and experience, and a big endeavor you'll need to support and keep improving and maintaining, you really want to nail the base user experience and not close options. Otherwise, you risk investing into the technical implementation and having to remove it as happened with the previous graph-based scripting system.

**Comment by Etheryans** [comment]
There's Hengo Script by Henrique Dart

https://hengoscript.com/
https://github.com/Henrique-Henrique/Hengo
https://x.com/henrique_dart/media
![Image](https://github.com/user-attachments/assets/2cb1bb5a-0b06-44bf-9756-63d5d7d6affe)

It seems like a great solution, talked to him on twitter but he believes that Godot's devs want to keep such tool separate, and so he does. Don't know if talk would be possible, but I wanted to leave this here :)

I'm not a programmer btw.

**Comment by Atenvardo** [comment]
I would love to see Hengo be the official Visual Scripting tool, especially for those learning as it compiles straight to GDScript which is visible at any stage, great for learning in a visual manner while easily seeing the generated code. So many great features with it, and more to come.

**Comment by Hypano** [comment]
I have a question / idea regarding the future direction of visual scripting in Godot.

Instead of building a fully separate visual scripting system, would it make sense to focus on a “visual layer on top of GDScript” rather than a replacement for it?

The idea would be:
- GDScript remains the single source of truth
- The visual system is just a representation layer (graph)
- Nodes are automatically generated from existing Godot APIs (ClassDB, methods, properties, signals)
- Custom functions written in GDScript automatically appear as nodes, with the ability to open each node and edit its underlying code.
- Complex logic can remain as code blocks instead of forcing everything into nodes

This would avoid the main problem VisualScript had before (missing features and duplication of the API), since it would always stay in sync with GDScript.

In addition, this approach could:
- Reduce maintenance effort (no separate scripting system)
- Leverage existing documentation directly
- Allow gradual adoption (users can mix visual and code freely)

Potentially, only small metadata additions would be very helpfull (for example: categories, node-friendly naming, or flags like “pure” vs “has side effects”).

Also: is Hengo already moving in this direction, or would this be considered a different approach?

I might be missing something here, so this could be a naive question,  but if this approach is feasible, is there a specific reason it hasn’t been explored more in the past?

I assume there are challenges (for example syncing between graph and code, or maintaining a good UX), but I would be interested to understand what the main blockers are.

**Comment by AThousandShips** [comment]
I don't see how this proposed solution would remove any of that duplication, what would "visual layer on top of" mean for the purposes of reducing those duplications

For comparison take a look at the system for visual shaders which seems to do what you propose here, and it is a large, hard to maintain system with a lot of code just to connect into the shader system

I would say that the best way to make visual scripting maintainable and manageable would be to integrate it either into the GDExtension system, or using an IR backend system like proposed in https://github.com/godotengine/godot-proposals/issues/8605

**Comment by Naros** [comment]
> * GDScript remains the single source of truth
> * The visual system is just a representation layer (graph)

It would actually be better to go the other way, because in our experience, users want to be able to organize their graphs according to their preferences, and using GDScript as the source of truth would imply that you'd need to store visual aspects separately or as comments, neither of which is an ideal way to present data that is logically coupled and it risks data integrity issues.

> * Custom functions written in GDScript automatically appear as nodes, with the ability to open each node and edit its underlying code.

What purpose then does a VS solution serve if you open a node and write GDScript directly? Sure, you could implement a sort of "expression" node like the one the VShader offers, but the vast majority of your users won't use that feature because their forte isn't development; they're designers and want designer-focused tools.

>  if this approach is feasible, is there a specific reason it hasn’t been explored more in the past?

Resources and the fact that the team, when releasing 4.0, conveyed this is best suited for the plugin community.

I'd rather see the Godot core devs focusing on more concrete components that make the engine better for the broader community, the GDScript devs focusing on ways to improve the language's feature set, performance, and innovation, with the idea that transpiling into GDScript becomes possible. This allows plugin authors to take advantage of all the low-level work the GDScript team does, while providing a higher-level experience for non-developers.

**Comment by Oniryans** [comment]
I'd like to come in, to weigh a bit with the perspective of somoeone having disabilities. I am autistic and have ADHD, but also more importantly have dyspraxia. While I can type on keyboards, doing so do pains my hands quickly, and it feels hard to me to code in traditional ways.
Visual scripting would ease that and make coding more accessible to persons having difficulties like mine. I feel that, as an accessibility feature, it may shouldn't be left aside completely.
I don't know how Hengo works, but it feels like a right balance that permits use of real GDscript in a visual manner, maybe it could be considered be made official?
I don't know if ressources permits it, but I'd still like to throw the idea out.

**Comment by AThousandShips** [comment]
Accessibility is absolutely important, but that can still be solved by plugins, and there are many areas of accessibility to be considered, and a lot of users who can't use visual scripting because of their specific accessibility needs, as nothing is a "silver bullet"

The addition of screen reader support is a potential avenue for some accessibility needs in this area as well (though probably not relevant in your case)

Maintaining a fourth language officially (GDScript, C#, and godot-cpp already being a thing, and technically we have 5 languages already if you count GDShader and visual shaders) would be a lot of overhead and work, and would mean that not only would it take resource from other parts of the engine, it would risk becoming neglected due to lack of demand

Which makes a plugin, potentially an officially endorsed one, much more manageable and well suited

**Comment by Oniryans** [comment]
I wouldn't mind an officially endorsed plugin, that'd be a great compromise

**Comment by Hypano** [comment]
Thanks for the replies, that helps me understand the topic a bit better.

My understanding is that visual shaders are still a separate, manually defined system, which is one reason why they are complex to maintain. What I had in mind is closer to generating nodes automatically from the existing API and documentation, so there would be less need to maintain a second representation of it. 
And not to add an additional language.

I also get the concern about storing visual layout. That is definitely a tradeoff I had not really considered.

I agree that the core team should focus on more important engine work. That is partly why I was thinking in the direction of a more automatic approach, even if the editor side would still be complex for the User.

This could maybe be extended through plugins later, for example to simplify certain workflows or provide more specific nodes. For the design side, small metadata additions could probably help a lot, for example categories, clearer node names, or flags for pure functions. Those might also be useful in other areas.

For me personally, this is mainly about reducing cognitive load. I can write code, but I do not always want to, and a visual layer could make writing code and reasoning about logic easier for me. The accessibility point mentioned above goes in a similar direction. Writing code is the biggest hurdle for me.

Appreciate the different perspectives.

**Comment by Naros** [comment]
@Oniryans feel free to check out [Orchestrator Visual Scripting](https://store.godotengine.org/asset/cratercrash-studios/orchestrator/) on the Asset store.

> What I had in mind is closer to generating nodes automatically from the existing API and documentation, so there would be less need to maintain a second representation of it.
And not to add an additional language.

This is precisely what Orchestrator does. It's heavily influenced by Unreal's Blueprint system.

**Comment by FR3NKD** [comment]
I made a small mockup on how I think the visual scripting solution should look like.
I think it's very important that it's 1:1 with GD Script and that it translates directly into GD Script like @Hypano was saying.

<img width="1161" height="842" alt="Image" src="https://github.com/user-attachments/assets/60f197a1-fcde-40cd-958b-96d1ad9c1cd5" />

**Comment by Hypano** [comment]
> [@Oniryans](https://github.com/Oniryans) feel free to check out [Orchestrator Visual Scripting](https://store.godotengine.org/asset/cratercrash-studios/orchestrator/) on the Asset store.
> 
> This is precisely what Orchestrator does. It's heavily influenced by Unreal's Blueprint system.

I checked out Orchestrator for a few hours now, and it is actually pretty cool. It is much closer to what I had in mind than I first thought.

But I also ran into the exact kind of problems that made me think about a GDScript-based visual layer in the first place. Some nodes were missing, and some things became much more complicated than they would be in code (for example, even something like converting an int to a string was not very straightforward).

In those situations, I would have liked to either create a small code node or open the underlying logic and write normal GDScript directly there. That would make it much more flexible, especially for quick checks or small changes without needing a dedicated node.

I know that mixing Orchestrator and GDScript is possible, but then the logic starts to live in multiple places, which can get hard to follow.

So I think the remaining difference is still the source of truth. Orchestrator already seems like a very good visual scripting system, but it still has some of the tradeoffs that come with having a separate underlying language.

Still, thanks for reminding me of that plugin again. I had read about it earlier in the thread, but did not give it much thought before your reminder.

**Comment by Naros** [comment]
> But I also ran into the exact kind of problems that made me think about a GDScript-based visual layer in the first place. Some nodes were missing, and some things became much more complicated than they would be in code (for example, even something like converting an int to a string was not very straightforward).

Was the `str(Variant)` function insufficient? 

> In those situations, I would have liked to either create a small code node or open the underlying logic and write normal GDScript directly there. That would make it much more flexible, especially for quick checks or small changes without needing a dedicated node.

While reasonable, the question then is how that GDScript gets turned into our bytecode understandable by our VM, while also being aware that that constant flip-flop across the GDExtension boundary is costly.

If one of the Godot devs have suggestions, we'd certainly be all ears.

**Comment by Oniryans** [comment]
> [@Oniryans](https://github.com/Oniryans) feel free to check out [Orchestrator Visual Scripting](https://store.godotengine.org/asset/cratercrash-studios/orchestrator/) on the Asset store.
> 
> > What I had in mind is closer to generating nodes automatically from the existing API and documentation, so there would be less need to maintain a second representation of it.
> > And not to add an additional language.
> 
> This is precisely what Orchestrator does. It's heavily influenced by Unreal's Blueprint system.

I don't know if Hengo generate automatically from the API, but it does generate real GDscript on it's side, instead of it's own separate language

**Comment by Oniryans** [comment]
Sorry for double post, but got a response from Henrique about Hengo:

<img width="1180" height="271" alt="Image" src="https://github.com/user-attachments/assets/69a1dd6d-ddf0-4ee4-a3d6-c51b25ec23a6" />

He talk about playmaker, but be aware that he took inspirations from many different visual script language from what I recall, blueprint being part of it (and this plugin do generate real Gdscript, inspired by the deprecated Bolt 2 for Unity that was supposed to generate C# directly)

**Comment by Hypano** [comment]
> Was the `str(Variant)` function insufficient?

@Naros I just tried it again and got it working pretty quickly this time. Not sure what went wrong before, maybe I missed something or it was related to the update ( i never installed) that was available shortly after I installed the plugin.


> > In those situations, I would have liked to either create a small code node or open the underlying logic and write normal GDScript directly there. That would make it much more flexible, especially for quick checks or small changes without needing a dedicated node.
> 
> While reasonable, the question then is how that GDScript gets turned into our bytecode understandable by our VM, while also being aware that that constant flip-flop across the GDExtension boundary is costly.

I get the concern about how it would be compiled or handled internally. That is actually what got me interested enough to build a small prototype over the last few days to explore the idea a bit further:
https://github.com/Hypano/GDScript-Visual-Layer

It is not meant as a finished system, but rather as a proof of concept to show what I had in mind. It simply edits the underlying GDScript file, so to speak. 



> I don't know if Hengo generate automatically from the API, but it does generate real GDscript on it's side, instead of it's own separate language

I also took a quick look at Hengo today. It felt a bit more complex at first glance, so I did not dive deep yet, but if it really generates GDScript instead of using a separate language, that sounds promising and I will take a closer look.

**Comment by Naros** [comment]
Hengo is effectively a GDScript generator with a visual front end. So yes, you do get the pros and cons of GDScript. It's a neat approach, but it raises questions when you apply it to larger projects with hundreds or thousands of scripts. Dual resource management is a real sore point for some users. And it's an open question how the approach scales to graphs with hundreds of nodes. The generator does linear scans through connection lists per port, which suggests real concerns at that size. Many Orchestrator users work with graphs of that scale, so it's something we've had to design for from the start.

Orchestrator took a different approach and was built as a first-class language alongside GDScript. This lets us avoid the dual-resource concern outright by allowing you to attach visual scripts directly to nodes, mirroring the behavior of the old Godot 3.x days. Orchestrator having its own language isn't a downside; it's a deliberate decision that lets us make our own choices about optimizations and features, and evolve independently of GDScript rather than be limited by its feature set. It's what's allowing us to introduce some of our current WIP features that wouldn't be possible otherwise.

The main reason I suspect developers prefer GDScript generators is having a fallback plan. When you build a serious project on open-source plugins, and the plugin ecosystem has a real staleness problem, developers want a recovery path if a plugin stops being maintained. So if the plugin goes defunct, at least you have the GDScript source, can learn GDScript, and continue forward. This is a concern we've heard from our own community, and one we sympathize with and agree with. It's why Orchestrator will soon ship with a GDScript exporter as well, so users get both: the runtime and extensibility advantages of a first-class language, plus a fallback escape hatch; they can graduate from visual to text-based programming, or mix and match as they want.
